packages/site/cms-site/src/graphQLFetch/graphQLFetch.ts

```diff
@@ -0,0 +1,86 @@
+import { SitePreviewParams } from "../sitePreview/SitePreviewApiHelper";
+
+type SitePreviewData = SitePreviewParams["settings"];
+type Fetch = typeof fetch;
+
+function graphQLHeaders(previewData?: SitePreviewData) {
+    const { includeInvisibleBlocks, includeInvisiblePages, previewDamUrls } = {
+        includeInvisiblePages: !!previewData,
+        includeInvisibleBlocks: previewData && previewData.includeInvisible,
+        previewDamUrls: !!previewData,
+    };
+
+    const headers: Record<string, string> = {};
+    const includeInvisibleContentHeaderEntries: string[] = [];
+    if (includeInvisiblePages) {
+        includeInvisibleContentHeaderEntries.push("Pages:Unpublished");
+    }
+    if (includeInvisibleBlocks) {
+        includeInvisibleContentHeaderEntries.push("Blocks:Invisible");
+    }
+    // tells api to send invisble content
+    // authentication is required when this header is used
+    if (includeInvisibleContentHeaderEntries.length > 0) {
+        headers["x-include-invisible-content"] = includeInvisibleContentHeaderEntries.join(",");
+    }
+    // tells api to create preview image urls
+    // authentication is required when this header is used
+    if (previewDamUrls) {
+        headers["x-preview-dam-urls"] = "1";
+    }
+    return headers;
+}
+
+//from graphql-request https://github.com/jasonkuhrt/graphql-request/blob/main/src/raw/functions/gql.ts
+export const gql = (chunks: TemplateStringsArray, ...variables: unknown[]): string => {
+    return chunks.reduce((acc, chunk, index) => `${acc}${chunk}${index in variables ? String(variables[index]) : ``}`, ``);
+};
+
+export function createFetchWithPreviewHeaders(fetch: Fetch, previewData?: SitePreviewData) {
+    const defaultHeaders = graphQLHeaders(previewData);
+    return createFetchWithDefaults(fetch, { headers: defaultHeaders });
+}
+
+export function createFetchWithDefaults(fetch: Fetch, defaults: RequestInit) {
+    return async function (input: RequestInfo, init?: RequestInit): Promise<Response> {
+        return fetch(input, {
+            ...defaults,
+            ...init,
+            headers: {
+                ...defaults.headers,
+                ...init?.headers,
+            },
+            next: {
+                ...defaults.next,
+                ...init?.next,
+            },
+        });
+    };
+}
+
+export type GraphQLFetch = <T, V>(query: string, variables?: V, init?: RequestInit) => Promise<T>;
+
+export function createGraphQLFetch(fetch: Fetch, url: string): GraphQLFetch {
+    return async function <T, V>(query: string, variables?: V, init?: RequestInit): Promise<T> {
+        const response = await fetch(url, {
+            method: "POST",
```

Comment:
Does this even work? Does Next cache POST requests?

https://github.com/vercel/next.js/blob/68de4c0357d83e9ee8c506b9d395404dd6e87483/packages/next/src/server/lib/patch-fetch.ts#L327-L339

Reply:
Yes it does :)

Reply:
what does _not_ work with POST is [request memoization](https://nextjs.org/docs/app/building-your-application/caching#opting-out) I just found in the docs. Don't know if that ever would be an issue.

Reply:
When having multiple server components that fetch the same GraphQL data in a single route it won't be memoized. But we can use [React's cache function](https://nextjs.org/docs/app/building-your-application/caching#react-cache-function) for this.

Reply:
Or use method GET